chia/data_layer/data_store.py

```diff
@@ -205,24 +209,25 @@ async def insert_row(self, table: bytes32, clvm_object: CLVMObject) -> TableRow:
         row_hash = sha256_treehash(sexp=clvm_object)
         clvm_bytes = SExp.to(clvm_object).as_bin()
 
-        await self.db.execute(
-            "INSERT INTO keys_values(key, value) VALUES(:key, :value)", {"key": row_hash.hex(), "value": clvm_bytes}
-        )
-
-        await self.db.execute(
-            "INSERT INTO table_values(table_id, key) VALUES(:table_id, :key)",
-            {"table_id": table.hex(), "key": row_hash.hex()},
-        )
+        async with self.db_wrapper.locked_transaction():
+            await self.db.execute(
+                "INSERT INTO keys_values(key, value) VALUES(:key, :value)", {"key": row_hash.hex(), "value": clvm_bytes}
+            )
 
-        await self.add_action(operation_type=OperationType.INSERT, key=row_hash, table=table)
+            await self.db.execute(
+                "INSERT INTO table_values(table_id, key) VALUES(:table_id, :key)",
+                {"table_id": table.hex(), "key": row_hash.hex()},
+            )
 
-        # TODO: Review reentrancy on .commit() since it isn't clearly tied to this
-        #       particular task's activity.
-        await self.db.commit()
+            await self._raw_add_action(operation_type=OperationType.INSERT, key=row_hash, table=table)
 
         return TableRow(clvm_object=clvm_object, hash=row_hash, bytes=clvm_bytes)
 
     async def add_action(self, operation_type: OperationType, key: bytes32, table: bytes32) -> None:
+        async with self.db_wrapper.locked_transaction():
+            return await self.add_action(operation_type=operation_type, key=key, table=table)
```

Comment:
is this a recursion ?

Reply:
Uh...  why yes, yes it is.

```suggestion
            return await self._raw_add_action(operation_type=operation_type, key=key, table=table)
```